### PR TITLE
feat(vpn): support some new parameters

### PR DESCRIPTION
**Pull request by Jason-Zhang9309**
**What this PR does / why we need it**:

**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
fixes #xxx

**Special notes for your reviewer**:

**Release note**:

```release-note
1. add new parameters network_type, er_id, access_vpc_id and access_subnet_id
2. support er in attachment_type
3. add a little improvement in RemoveNil func
4. update the doc and test cases
```

## PR Checklist

* [x] Tests added/passed.
* [x] Documentation updated.
* [x] Schema updated.

## Acceptance Steps Performed

```
make testacc TEST='./huaweicloud/services/acceptance/vpn' TESTARGS='-run TestAccGateway_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/vpn -v -run TestAccGateway_basic -timeout 360m -parallel 4
=== RUN   TestAccGateway_basic
=== PAUSE TestAccGateway_basic
=== CONT  TestAccGateway_basic
--- PASS: TestAccGateway_basic (173.72s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/vpn       173.810s

make testacc TEST='./huaweicloud/services/acceptance/vpn' TESTARGS='-run TestAccGateway_withER'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/vpn -v -run TestAccGateway_withER -timeout 360m -parallel 4
=== RUN   TestAccGateway_withER
=== PAUSE TestAccGateway_withER
=== CONT  TestAccGateway_withER
--- PASS: TestAccGateway_withER (251.80s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/vpn       251.905s
```

### CODE REVIEW

File: docs/resources/vpn_gateway.md [hunk 1]
```diff
@@ -75,37 +94,62 @@ The following arguments are supported:
...
 
-* `local_subnets` - (Required, List) The list of local subnets.
+* `flavor` - (Optional, String, ForceNew) The flavor of the VPN gateway.
+  The value can be **Basic**, **Professional1** and **Professional2**. Defaults to **Professional1**.
+
+  Changing this parameter will create a new resource.
 
-* `connect_subnet` - (Required, String, ForceNew) The Network ID of the VPC subnet used by the VPN gateway.
+* `attachment_type` - (Optional, String, ForceNew) The attachment type. The value can be **vpc** and **er**.
+  Defaults to **vpc**
```
Comment (houpeng80) [inline docs/resources/vpn_gateway.md]:
```suggestion
  Defaults to **vpc**.
```

File: docs/resources/vpn_gateway.md [hunk 2]
```diff
@@ -75,37 +94,62 @@ The following arguments are supported:
...
 
   Changing this parameter will create a new resource.
 
-* `availability_zones` - (Required, List, ForceNew) The list of availability zone IDs.
+* `network_type` - (Optional, String, ForceNew) The network type. The value can be **public** and **private**.
+  Defaults to **public**
```
Comment (houpeng80) [inline docs/resources/vpn_gateway.md]:
```suggestion
  Defaults to **public**.
```

File: docs/resources/vpn_gateway.md [hunk 3]
```diff
@@ -133,8 +177,9 @@ The `master_eip` or `slave_eip` block supports:
 
   Changing this parameter will create a new resource.
 
-* `bandwidth_size` - (Optional, Int, ForceNew) Bandwidth size in Mbit/s. When the `flavor` is **V300**, the value
-  cannot be greater than **300**. When the `flavor` is **V1G**, the value cannot be greater than **1024**.
+* `bandwidth_size` - (Optional, Int, ForceNew) Bandwidth size in Mbit/s. When the `flavor` is **Basic**, the value
+  cannot be greater than **100**. When the `flavor` is **Professional1**, the value cannot be greater than **300**.
+  When the `flavor` is **Professional2**, the value cannot be greater than **1024**.
```
Comment (houpeng80) [inline docs/resources/vpn_gateway.md]:
```suggestion
  When the `flavor` is **Professional2**, the value cannot be greater than **1000**.
```

File: huaweicloud/services/acceptance/vpn/resource_huaweicloud_vpn_gateway_test.go [hunk 4]
```diff
@@ -176,3 +227,45 @@ resource "huaweicloud_vpn_gateway" "test" {
...
+
+data "huaweicloud_availability_zones" "test" {}
+
+resource "huaweicloud_er_instance" "test" {
+  availability_zones = [
+    data.huaweicloud_availability_zones.test.names[0],
+    data.huaweicloud_availability_zones.test.names[3]
+  ]
+
+  name = "zhangjishu-test"
```
Comment (houpeng80) [inline huaweicloud/services/acceptance/vpn/resource_huaweicloud_vpn_gateway_test.go]: how about replaced with an environment variable?

File: docs/resources/vpn_gateway.md (hunk 1 above)
Comment (Jason-Zhang9309) [reply docs/resources/vpn_gateway.md]: fixed

File: docs/resources/vpn_gateway.md (hunk 2 above)
Comment (Jason-Zhang9309) [reply docs/resources/vpn_gateway.md]: fixed

File: huaweicloud/services/acceptance/vpn/resource_huaweicloud_vpn_gateway_test.go (hunk 4 above)
Comment (Jason-Zhang9309) [reply huaweicloud/services/acceptance/vpn/resource_huaweicloud_vpn_gateway_test.go]: fixed

File: docs/resources/vpn_gateway.md (hunk 3 above)
Comment (Jason-Zhang9309) [reply docs/resources/vpn_gateway.md]: fixed

**Comment by Jason-Zhang9309** [comment]
```
make testacc TEST='./huaweicloud/services/acceptance/vpn' TESTARGS='-run TestAccGateway_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/vpn -v -run TestAccGateway_basic -timeout 360m -parallel 4
=== RUN   TestAccGateway_basic
=== PAUSE TestAccGateway_basic
=== CONT  TestAccGateway_basic
--- PASS: TestAccGateway_basic (164.03s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/vpn       164.101s

make testacc TEST='./huaweicloud/services/acceptance/vpn' TESTARGS='-run TestAccGateway_withER'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/vpn -v -run TestAccGateway_withER -timeout 360m -parallel 4
=== RUN   TestAccGateway_withER
=== PAUSE TestAccGateway_withER
=== CONT  TestAccGateway_withER
--- PASS: TestAccGateway_withER (245.87s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/vpn       245.941s
```

File: docs/resources/vpn_gateway.md [hunk 5]
```diff
@@ -75,37 +94,62 @@ The following arguments are supported:
 
 * `name` - (Required, String) The name of the VPN gateway. Only letters, digits, underscores(_) and hypens(-) are supported.
 
-* `vpc_id` - (Required, String, ForceNew) The ID of the VPC to which the VPN gateway is connected.
+* `availability_zones` - (Required, List, ForceNew) The list of availability zone IDs.
```
Comment (ShiChangkuo) [inline docs/resources/vpn_gateway.md:97]: it's better to change the order in schema

File: huaweicloud/services/vpn/resource_huaweicloud_vpn_gateway.go [hunk 6]
```diff
@@ -87,15 +90,45 @@ func ResourceGateway() *schema.Resource {
...
+				ForceNew:    true,
+				Computed:    true,
+				Description: `The access subnet ID of the VPN gateway.`,
+			},
+			"er_id": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				ForceNew:    true,
+				Computed:    true,
+				Description: `The enterprise router ID to attach with to VPN gateway..`,
```
Comment (ShiChangkuo) [inline huaweicloud/services/vpn/resource_huaweicloud_vpn_gateway.go]: ending with double `.`
```suggestion
				Description: `The enterprise router ID to attach with to VPN gateway.`,
```

Comment (Jason-Zhang9309) [reply huaweicloud/services/vpn/resource_huaweicloud_vpn_gateway.go]: fixed

File: docs/resources/vpn_gateway.md (hunk 5 above)
Comment (Jason-Zhang9309) [reply docs/resources/vpn_gateway.md:97]: sure, the order is updated

**Pull request merged**